Source/Core/DolphinQt/DiscordJoinRequestDialog.cpp

```diff
@@ -0,0 +1,96 @@
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#ifdef USE_DISCORD_PRESENCE
+
+#include <QGridLayout>
+#include <QLabel>
+#include <QPixmap>
+#include <QPushButton>
+
+#include <discord-rpc/include/discord_rpc.h>
+
+#include "Common/HttpRequest.h"
+#include "Common/StringUtil.h"
+
+#include "DolphinQt/DiscordJoinRequestDialog.h"
+
+DiscordJoinRequestDialog::DiscordJoinRequestDialog(QWidget* parent, const char* id,
+                                                   const std::string& discord_tag,
+                                                   const std::string& avatar)
+    : QDialog(parent), m_user_id(id), m_close_timestamp(std::time(nullptr) + s_max_lifetime_seconds)
+{
+  setWindowTitle(tr("Request to Join Your Party"));
+  setWindowFlags(windowFlags() & ~Qt::WindowContextHelpButtonHint);
+
+  QPixmap avatar_pixmap;
+
+  if (avatar.empty())
+  {
+    const std::string avatar_endpoint =
+        StringFromFormat("https://cdn.discordapp.com/avatars/%s/%s.png", id, avatar.c_str());
+
+    Common::HttpRequest request;
+    Common::HttpRequest::Response response = request.Get(avatar_endpoint);
+
+    if (response.has_value())
+      avatar_pixmap.loadFromData(response->data(), static_cast<uint>(response->size()), "png");
+  }
+
+  CreateLayout(discord_tag, avatar_pixmap);
+  ConnectWidgets();
+}
+
+std::time_t DiscordJoinRequestDialog::GetCloseTimestamp() const
+{
+  return m_close_timestamp;
+}
+
+void DiscordJoinRequestDialog::CreateLayout(const std::string& discord_tag, const QPixmap& avatar)
+{
+  m_main_layout = new QGridLayout;
+
+  QString invite_button_text = QString::fromWCharArray(L"\u2714 ");
+  QString decline_button_text = QString::fromUcs4(U"\x1F5D9 ");
+  invite_button_text += tr("Invite");
+  decline_button_text += tr("Decline");
+
+  m_invite_button = new QPushButton(invite_button_text);
+  m_decline_button = new QPushButton(decline_button_text);
+  m_ignore_button = new QPushButton(tr("Ignore"));
+
+  QLabel* text =
+      new QLabel(tr("%1\nwants to join your party.").arg(QString::fromStdString(discord_tag)));
+  text->setAlignment(Qt::AlignCenter);
+
+  if (!avatar.isNull())
+  {
+    QLabel* picture = new QLabel();
+    picture->setPixmap(avatar);
+    m_main_layout->addWidget(picture, 1, 0, 1, 3, Qt::AlignHCenter);
+  }
+
+  m_main_layout->addWidget(text, 2, 0, 3, 3, Qt::AlignHCenter);
+  m_main_layout->addWidget(m_invite_button, 8, 0);
+  m_main_layout->addWidget(m_decline_button, 8, 1);
+  m_main_layout->addWidget(m_ignore_button, 8, 2);
+
+  setLayout(m_main_layout);
+}
+
+void DiscordJoinRequestDialog::ConnectWidgets()
+{
+  connect(m_invite_button, &QPushButton::pressed, [this] { Reply(DISCORD_REPLY_YES); });
+  connect(m_decline_button, &QPushButton::pressed, [this] { Reply(DISCORD_REPLY_NO); });
+  connect(m_ignore_button, &QPushButton::pressed, [this] { Reply(DISCORD_REPLY_IGNORE); });
+  connect(this, &QDialog::rejected, this, [this] { Reply(DISCORD_REPLY_IGNORE); });
```

Comment:
What exactly does "ignore" do? Should we really have closing the dialog perform the "ignore" action, or just the "no" action?

Reply:
Ah okay, I was thinking the ignore button would block future requests. Thanks for the clarification. Ignore does seem like the appropriate action for closing the dialog in that case.